### PR TITLE
Add the remaining SIMD ops

### PR DESCRIPTION
**Pull request by justinmichaud**
```
[SIMD] Add SIMD opcodes and spec tests
https://bugs.webkit.org/show_bug.cgi?id=246343
rdar://problem/101295960

Reviewed by NOBODY (OOPS!).

- Implement v128 globals support
    - This unfortunately doubles the space occupied by every wasm global
- Import WASM simd spec tests plus a few V8 tests
    - These (will be OOPS) skipped for now because they don't all pass
- Add an implementation for every intruction on ARM
- Add a debugging flag to dump WASM binaries

To quickly test, run:

Tools/Scripts/run-jsc-stress-tests --jsc `pwd`/WebKitBuild/Debug/jsc JSTests/wasm.yaml --filter ".*(simd|wasm/v8/exceptions|wasm/stress/mutable-globals.js|wasm/stress/immutable-globals.js|wasm/stress/i64-extract.js|wasm/spec-tests/exports.wast.js|wasm/function-references-spec-tests/call_ref.wast.js).*"

Currently we pass 75% of the new tests.```

### CODE REVIEW

File: JSTests/wasm.yaml [hunk 1]
```diff
@@ -48,7 +48,7 @@
 - path: wasm/self-test/
   cmd: runWebAssemblySuite unless parseRunCommands
 - path: wasm/v8/
-  cmd: runWebAssemblySuite(:no_module, "mjsunit.js") unless parseRunCommands
+  cmd: runWebAssemblySuite(:no_module, "mjsunit.js", "--useWebAssemblySIMD=1") unless parseRunCommands
```
Comment (justinmichaud) [inline JSTests/wasm.yaml:51]: note that this technically reduces our test coverage?

File: Source/JavaScriptCore/wasm/WasmStreamingParser.cpp [hunk 2]
```diff
@@ -69,11 +70,36 @@ NEVER_INLINE auto WARN_UNUSED_RETURN StreamingParser::fail(Args... args) -> Stat
     return State::FatalError;
 }
 
+static void dumpWasmSource(const Vector<uint8_t>& source)
+{
+    static int count = 0;
+    const char* file = Options::dumpWasmSourceFileName();
+    if (!file)
+        return;
+    auto fileHandle = FileSystem::openFile(WTF::makeString(file) + (count++) + ".wasm"_s,
```
Comment (Constellation) [inline Source/JavaScriptCore/wasm/WasmStreamingParser.cpp]: Use `makeString(file, count++, ".wasm"_s)`.

File: Source/JavaScriptCore/wasm/WasmStreamingParser.cpp [hunk 3]
```diff
@@ -69,11 +70,36 @@ NEVER_INLINE auto WARN_UNUSED_RETURN StreamingParser::fail(Args... args) -> Stat
...
+        return;
+    auto fileHandle = FileSystem::openFile(WTF::makeString(file) + (count++) + ".wasm"_s, 
+        FileSystem::FileOpenMode::Write,
+        FileSystem::FileAccessPermission::All,
+        /* failIfFileExists = */ true);
+    if (fileHandle == FileSystem::invalidPlatformFileHandle) {
+        dataLogLn("Error dumping wasm");
+        return;
+    }
+    dataLogLn("Dumping ", source.size(), " wasm source bytes to ", WTF::makeString(file), (count - 1), ".wasm");
```
Comment (Constellation) [inline Source/JavaScriptCore/wasm/WasmStreamingParser.cpp]: Ditto.

File: Source/JavaScriptCore/wasm/WasmSectionParser.cpp [hunk 4]
```diff
@@ -639,6 +649,17 @@ auto SectionParser::parseInitExpr(uint8_t& opcode, uint64_t& bitsOrImportNumber,
         break;
     }
 
+    case ExtSIMD: {
+        WASM_PARSER_FAIL_IF(!parseUInt8(opcode), "can't get init_expr's simd opcode");
+        WASM_PARSER_FAIL_IF((ExtSIMDOpType) opcode != ExtSIMDOpType::V128Const, "unknown init_expr simd opcode ", opcode);
```
Comment (Constellation) [inline Source/JavaScriptCore/wasm/WasmSectionParser.cpp]: Use `static_cast`

File: Source/JavaScriptCore/wasm/WasmSectionParser.cpp [hunk 5]
```diff
@@ -639,6 +649,17 @@ auto SectionParser::parseInitExpr(uint8_t& opcode, uint64_t& bitsOrImportNumber,
...
+        v128_t constant;
```
Comment (Constellation) [inline Source/JavaScriptCore/wasm/WasmSectionParser.cpp:656]: We should check `Options::useWebAssemblySIMD`.

File: Source/JavaScriptCore/wasm/WasmGlobal.h [hunk 6]
```diff
@@ -43,16 +43,25 @@ class Global final : public ThreadSafeRefCounted<Global> {
     WTF_MAKE_FAST_ALLOCATED(Global);
 public:
     union Value {
+        v128_t m_vector { };
         uint64_t m_primitive;
         WriteBarrierBase<Unknown> m_externref;
         Value* m_pointer;
+
+        Value() = default;
```
Comment (Constellation) [inline Source/JavaScriptCore/wasm/WasmGlobal.h]: Why?

File: Source/JavaScriptCore/wasm/WasmGlobal.cpp [hunk 7]
```diff
@@ -87,6 +93,10 @@ void Global::set(JSGlobalObject* globalObject, JSValue argument)
         m_value.m_primitive = bitwise_cast<uint64_t>(value);
         break;
     }
+    case TypeKind::V128: {
+        throwException(globalObject, throwScope, createJSWebAssemblyRuntimeError(globalObject, vm, "Cannot set value of v128 global"_s));
```
Comment (Constellation) [inline Source/JavaScriptCore/wasm/WasmGlobal.cpp:97]: return here.

File: Source/JavaScriptCore/wasm/WasmGlobal.cpp [hunk 8]
```diff
@@ -38,15 +38,21 @@ namespace JSC { namespace Wasm {
...
     case TypeKind::I64:
-        return JSBigInt::makeHeapBigIntOrBigInt32(globalObject, static_cast<int64_t>(m_value.m_primitive));
+        RELEASE_AND_RETURN(throwScope, JSBigInt::makeHeapBigIntOrBigInt32(globalObject, static_cast<int64_t>(m_value.m_primitive)));
     case TypeKind::F32:
         return jsNumber(purifyNaN(static_cast<double>(bitwise_cast<float>(static_cast<uint32_t>(m_value.m_primitive)))));
     case TypeKind::F64:
         return jsNumber(purifyNaN(bitwise_cast<double>(m_value.m_primitive)));
+    case TypeKind::V128:
+        throwException(globalObject, throwScope, createJSWebAssemblyRuntimeError(globalObject, vm, "Cannot get value of v128 global"_s));
+        return jsUndefined();
```
Comment (Constellation) [inline Source/JavaScriptCore/wasm/WasmGlobal.cpp]: We use `return { };` usually.

File: Source/JavaScriptCore/b3/air/AirOpcode.opcodes [hunk 9]
```diff
@@ -107,6 +107,8 @@
 
 Nop
 
+arm64: Breakpoint
```
Comment (Constellation) [inline Source/JavaScriptCore/b3/air/AirOpcode.opcodes]: Why not just making it non-arm64 only?

**Pull request closed without merging**